### PR TITLE
Pass options via a json file instead of command line arguments

### LINKED ISSUE #164: [MLH] Pass options via a json file instead of command line arguments
Mariana Trench consists of:
* The shim, a small python script, which parses user-provided command line options and calls the binary
* The C++ binary, which actually performs the analysis.

To pass the user-provided options from python to C++, we currently build a very long command from python:
https://github.com/facebook/mariana-trench/blob/main/shim/shim.py#L610-L748
We then parse those options from C++:
https://github.com/facebook/mariana-trench/blob/main/source/Options.cpp#L151-L306

While this works fine, we risk running into the maximum size of a command line (although unlikely): https://unix.stackexchange.com/questions/120642/what-defines-the-maximum-size-for-a-command-single-argument
It is also painful to debug problems with argument passing.

For these reasons, we would like to implement a new way of passing arguments:
The shim should create a temporary json file containing all parameters, and the C++ binary would just parse that file.

### PR DESCRIPTION
**Pull request by ZeyadTarekk**
Fixes #164 

- First, the shim parses the arguments and creates a JSON file having those arguments
- Then the shim passes the JSON file path to the cpp
- The CPP code parses that JSON file and uses the arguments

### CODE REVIEW

File: shim/shim.py [hunk 1]
```diff
@@ -755,6 +755,154 @@ def _get_command_options(
     return options
 
 
+def _get_command_options_json(
+    arguments: argparse.Namespace, apk_directory: str, dex_directory: str
+) -> str:
+
```
Comment (arthaud) [inline shim/shim.py:761]: please follow our coding conventions. We don't add extra newlines at the beginning of functions.

File: shim/shim.py [hunk 2]
```diff
@@ -755,6 +755,154 @@ def _get_command_options(
...
+    if arguments.job_id:
+        options_json["job-id"] = arguments.job_id
+
+    if arguments.metarun_id:
+        options_json["metarun-id"] = arguments.metarun_id
+
+    trace_settings = [f"MARIANA_TRENCH:{arguments.verbosity}"]
+    if "TRACE" in os.environ:
+        trace_settings.insert(0, os.environ["TRACE"])
+    os.environ["TRACE"] = ",".join(trace_settings)
```
Comment (arthaud) [inline shim/shim.py:864]: This is not really related to the options file, can we move this in a different function, for instance `set_environment_variables(arguments)` or something?

File: shim/shim.py [hunk 3]
```diff
@@ -755,6 +755,154 @@ def _get_command_options(
...
+    if arguments.always_export_origins:
+        options_json["always-export-origins"] = arguments.always_export_origins
+
+    # Dump the options to a file and return the file path
+    options_file = tempfile.NamedTemporaryFile(suffix=".json",delete=False)
+    options_file.write(json.dumps(options_json).encode())
+    # Get the options file path
+    options_file_path = options_file.name
+
+    return options_file_path
```
Comment (arthaud) [inline shim/shim.py:903]: unnecessary temporary variable
```suggestion
    return options_file.name
```

File: source/Main.cpp [hunk 4]
```diff
@@ -40,11 +42,43 @@ int main(int argc, char* argv[]) {
   tool.add_options(options);
 
   try {
+    if (argc < 2) {
+      std::cerr << "Usage: " << argv[0] << " <json_config_file>\n";
+      return ExitCode::invalid_argument_error("No JSON configuration file provided.");
+    }
```
Comment (arthaud) [inline source/Main.cpp:48]: We should use boost program options to get the file name, instead of parsing the argument ourselves.
See https://www.boost.org/doc/libs/1_58_0/doc/html/program_options/tutorial.html#idp337609360
We need to use `positional_options_description`

File: source/Main.cpp [hunk 5]
```diff
@@ -40,11 +42,43 @@ int main(int argc, char* argv[]) {
...
+    Json::Value json = marianatrench::JsonReader::parse_json_file(json_file_path);
+    // Validate the JSON object
+    marianatrench::JsonValidation::validate_object(json);
+    // Use variables_map to store the JSON data
     program_options::variables_map variables;
-    program_options::store(
-        program_options::command_line_parser(argc, argv).options(options).run(),
-        variables);
+    // Populate variables_map from JSON data
+    for (const auto& key : json.getMemberNames()) {
```
Comment (arthaud) [inline source/Main.cpp:59]: While this work, this is less than ideal. 

What we want instead is to have a static method `Options::from_json_file()`, without using the intermediate `program_options::variables_map`

File: shim/shim.py [hunk 6]
```diff
@@ -612,147 +612,148 @@ def _add_debug_arguments(parser: argparse.ArgumentParser) -> None:
...
+            options_json["allow-via-cast-feature"].append(feature.strip())
 
     if arguments.heuristics:
-        options.append("--heuristics")
-        options.append(arguments.heuristics)
+        options_json["heuristics"] = arguments.heuristics
 
     if arguments.sequential:
-        options.append("--sequential")
+        options_json["sequential"] = arguments.sequential
```
Comment (arthaud) [inline shim/shim.py:687]: nitpick here and a few other cases: if we entered that `if` then we know `arguments.sequential` is True, so let's just write:
```suggestion
        options_json["sequential"] = True
```

File: shim/shim.py [hunk 7]
```diff
@@ -612,147 +612,148 @@ def _add_debug_arguments(parser: argparse.ArgumentParser) -> None:
...
+        options_json["maximum-method-analysis-time"] = arguments.maximum_method_analysis_time
+
     if arguments.enable_cross_component_analysis:
-        options.append("--enable-cross-component-analysis")
+        options_json["enable-cross-component-analysis"] = arguments.enable_cross_component_analysis
+
     if arguments.extra_analysis_arguments:
-        options.extend(shlex.split(arguments.extra_analysis_arguments))
+        for arg in shlex.split(arguments.extra_analysis_arguments):
+            options_json[arg] = arg
```
Comment (arthaud) [inline shim/shim.py:715]: arg, I forgot about this. We used this to inject extra parameters for the C++ binary.
I don't think your implementation works, for instance we would be calling it like `--extra-analysis-arguments=--enable-cross-component-analysis`.
I propose to let users pass arbitrary json and merge it with the options. Something like:
```
if arguments.extra_analysis_arguments:
   extra_arguments = json.loads(arguments.extra_analysis_arguments)
   for key, value in extra_arguments.items():
     if key in options_json and isinstance(options_json[key], list) and isinstance(value, list):
       # append to the existing options
       options_json[key].extend(value)
     else:
      # override
      options_json[key] = value
```
This should also be moved towards the end of this function

File: shim/shim.py [hunk 8]
```diff
@@ -612,147 +612,148 @@ def _add_debug_arguments(parser: argparse.ArgumentParser) -> None:
...
     if arguments.metarun_id:
-        options.append("--metarun-id")
-        options.append(arguments.metarun_id)
+        options_json["metarun-id"] = arguments.metarun_id
 
-    trace_settings = [f"MARIANA_TRENCH:{arguments.verbosity}"]
-    if "TRACE" in os.environ:
-        trace_settings.insert(0, os.environ["TRACE"])
-    os.environ["TRACE"] = ",".join(trace_settings)
+    _set_environment_variables(arguments)
```
Comment (arthaud) [inline shim/shim.py:723]: I would call this directly from `main`. It shouldn't be in `_get_command_options_json` since it has nothing to do with the json options.

File: shim/shim.py [hunk 9]
```diff
@@ -855,14 +856,19 @@ def main() -> None:
                 binary, arguments, apk_directory, dex_directory
```
Comment (arthaud) [inline shim/shim.py:856]: damn it, this also needs to be changed, but it's facebook-internal code. I will need to change it myself before merging the PR.

File: shim/shim.py [hunk 10]
```diff
@@ -855,14 +856,19 @@ def main() -> None:
...
-            if arguments.gdb:
-                command = ["gdb", "--args"] + command
-            elif arguments.lldb:
-                command = ["lldb", "--"] + command
-            LOG.info(f"Running Mariana Trench: {' '.join(command)}")
-            output = subprocess.run(command)
+            with tempfile.NamedTemporaryFile(suffix=".json",mode="w") as options_file:
+                options_json = _get_command_options_json(arguments, apk_directory, dex_directory)
+                print(f"options_json type: {type(options_json)}")
+                print(f"options_file mode: {options_file.mode}")
```
Comment (arthaud) [inline shim/shim.py:862]: Please remove those debugging statements

File: source/Main.cpp [hunk 11]
```diff
@@ -34,23 +36,26 @@ int main(int argc, char* argv[]) {
...
       return 0;
     }
-
-    program_options::notify(variables);
+    if (!variables.count("config")) {
+      std::cerr << "No JSON configuration file provided.\n";
+      std::cerr << "Usage: " << argv[0] << " --config <json_config_file>\n";
+      return ExitCode::invalid_argument_error("No JSON configuration file provided.");
+    }
+
```
Comment (arthaud) [inline source/Main.cpp:58]: Is this necessary? I think boost should already error when calling `program_options::parse_command_line` since you marked `config` as required.

File: source/MarianaTrench.cpp [hunk 12]
```diff
@@ -423,8 +423,9 @@ std::vector<std::string> filter_existing_jars(
 
 void MarianaTrench::run(const program_options::variables_map& variables) {
   Context context;
-
-  context.options = std::make_unique<Options>(variables);
+  std::string json_file_path = variables["config"].as<std::string>();
+  std::unique_ptr<Options> options_from_file = Options::options_from_json_file(json_file_path);
+  context.options = std::move(options_from_file);
```
Comment (arthaud) [inline source/MarianaTrench.cpp:428]: you can avoid the temporary variable:
```suggestion
  context.options = Options::options_from_json_file(json_file_path);
```

File: source/Options.cpp [hunk 13]
```diff
@@ -303,6 +303,38 @@ Options::Options(const boost::program_options::variables_map& variables) {
...
+      } else if (value.isArray()) {
+        std::vector<std::string> array_values;
+        for (const auto& element : value) {
+          if (element.isString()) {
+            array_values.push_back(element.asString());
+          }
+        }
+        variables.insert(std::make_pair(key, program_options::variable_value(array_values, false)));
+      }
+    }
```
Comment (arthaud) [inline source/Options.cpp:335]: We should change this to parse the json and update options directly like in the method above, without using `program_options::variables_map`.
Ultimately we should remove those 2 methods:
```
explicit Options(const boost::program_options::variables_map& variables);
static void add_options(boost::program_options::options_description& options);
```

**Review by arthaud — CHANGES_REQUESTED** [review]

File: source/Main.cpp (hunk 11 above)
Comment (ZeyadTarekk) [reply source/Main.cpp:58]: It is necessary, I tried to remove the `--config` from the Python shim and it didn't give an error.

File: source/Options.cpp [hunk 14]
```diff
@@ -147,336 +147,176 @@ Options::Options(
...
-Options::Options(const boost::program_options::variables_map& variables) {
-  system_jar_paths_ = parse_paths_list(
-      variables["system-jar-paths"].as<std::string>(),
-      std::nullopt,
-      /* check exist */ false);
+Options::Options(const Json::Value &json){
+    system_jar_paths_ = parse_paths_list(
+        json["system-jar-paths"].asString(),
+        std::nullopt,
+        /* check exist */ false);
```
Comment (arthaud) [inline source/Options.cpp:154]: using `json["system-jar-paths"].asString()` converts the list to a string and `parse_paths_list` converts the string back to a list. This is wasteful. We should write a `JsonValidation::string_list()` function. See JsonValidation.cpp for examples on how to parse a list of json elements.

File: source/Options.cpp [hunk 15]
```diff
@@ -147,336 +147,176 @@ Options::Options(
...
 
-  apk_directory_ =
-      check_directory_exists(variables["apk-directory"].as<std::string>());
-  dex_directory_ =
-      check_directory_exists(variables["dex-directory"].as<std::string>());
+    apk_directory_ =
+        check_directory_exists(json["apk-directory"].asString());
```
Comment (arthaud) [inline source/Options.cpp:157]: Here and many other places: you can use JsonValidation::string like this:
```suggestion
    apk_directory_ =
        check_directory_exists(JsonValidation::string(json, "apk-directory"));
```

File: source/Options.cpp [hunk 16]
```diff
@@ -147,336 +147,176 @@ Options::Options(
...
+            json["sharded-models-directory"].asString()));
+    }
 
-  if (!variables["source-exclude-directories"].empty()) {
-    source_exclude_directories_ = parse_paths_list(
-        variables["source-exclude-directories"].as<std::string>(),
-        /* extension */ std::nullopt,
-        /* check_exist */ false);
-  }
+    sequential_ = json["sequential"].asBool();
```
Comment (arthaud) [inline source/Options.cpp:248]: When `--sequential` is not provided to the command line in the shim, we don't even define a `sequential` key in json (see your code in the shim). In that case, I think `json["sequential"]` will be null. I'm surprised `asBool()` doesn't error on that.
Can we define a function `JsonValidation::optional_boolean(json, key, default_value)` and use it here and for the other booleans?

File: source/Options.cpp [hunk 17]
```diff
@@ -147,336 +147,176 @@ Options::Options(
...
 
-  if (!variables["grepo-metadata-path"].empty()) {
-    grepo_metadata_path_ =
-        check_path_exists(variables["grepo-metadata-path"].as<std::string>());
-  }
+    if (!json["log-method"].isNull()) {
+        for (const auto& value : json["log-method"]) {
+            log_methods_.push_back(value.asString());
+        }
+    }
```
Comment (arthaud) [inline source/Options.cpp:275]: * Instead of `json[key].isNull()` prefer using `json.isMember(key)`
* This will crash if `json["log-method"]` is not an array. Use `JsonValidation::nonempty_array()`
* Use `JsonValidation::string` instead of `.asString()` because `.asString()` throws exceptions that we don't handle, but `JsonValidation::string` throws proper exceptions that we catch and provide to the user.

```suggestion
    if (json.isMember("log-method")) {
        for (const auto& value : JsonValidation::nonempty_array(json, "log-method")) {
            log_methods_.push_back(JsonValidation::string(value));
        }
    }
```

**Review by arthaud — CHANGES_REQUESTED** [review]

File: source/Options.cpp (hunk 14 above)
Comment (ZeyadTarekk) [reply source/Options.cpp:154]: Here We should use `parse_paths_list ` as it handles if there is a directory if extracts the files on it

Comment (anwesht) [reply source/Options.cpp]: I think what @arthaud means is: 
- Now that we are passing the arguments as a json file, we should take arguments that take a list as a json array and *not* a `;` separated string like before. 
- For eg. we want to pass 2 paths to `system-jar-paths` argument: `path/to/1` and `path/to/2`.
- You are creating the options json as: 
```
  "system-jar-paths": "path/to/1;path/to/2"
```
- In `shim.py`, you already split this argument by `;` to check if the paths are valid. So no need to join the list back into a string anymore since we are creating a json file. 
  - See https://github.com/ZeyadTarekk/mariana-trench/blob/3401b4818bcc14c2ea8f3d74c89271540b539860/shim/shim.py#L69-L75. You'll have to update that and other similar helpers in `shim.py` accordingly.

- Instead of the string, we should create a json array for these keys like:
```
"system-jar-paths": ["path/1", "path/2"]
```
- Then, here you can add a `JsonValidation::string_list()` method to parse out this json array 
- `parse_paths_list()` currently takes a string and splits it on `;`. This then will not be necessary.

**Review by anwesht — CHANGES_REQUESTED** [review]

File: source/Options.cpp [hunk 18]
```diff
@@ -147,336 +147,164 @@ Options::Options(
...
-      "Enable taint propagation across object type arguments. By default, taint propagation is only tracked for return values and the `this` argument. This enables taint propagation across method invocations for all other object type arguments as well.");
-  options.add_options()(
-      "heuristics",
-      program_options::value<std::string>(),
-      "Path to JSON configuration file which specifies heuristics parameters to use during the analysis. See the documentation for available heuristics parameters.");
+
+std::unique_ptr<Options> Options::options_from_json_file(const std::string& options_json_path){
+    // Use JsonReader to parse the JSON file
+    Json::Value json = marianatrench::JsonReader::parse_json_file(options_json_path);
+    // Validate the JSON object
```
Comment (arthaud) [inline source/Options.cpp:305]: nitpick: would remove those comments, since the function are pretty well named, this is not adding much information.

File: source/Options.cpp [hunk 19]
```diff
@@ -147,336 +147,164 @@ Options::Options(
...
-      "Export origin information for all frames instead of only leaves. Used for debugging.");
-  options.add_options()(
-      "propagate-across-arguments",
-      "Enable taint propagation across object type arguments. By default, taint propagation is only tracked for return values and the `this` argument. This enables taint propagation across method invocations for all other object type arguments as well.");
-  options.add_options()(
-      "heuristics",
-      program_options::value<std::string>(),
-      "Path to JSON configuration file which specifies heuristics parameters to use during the analysis. See the documentation for available heuristics parameters.");
+
+std::unique_ptr<Options> Options::options_from_json_file(const std::string& options_json_path){
```
Comment (arthaud) [inline source/Options.cpp:302]: nitpick: would remove the additional `options_` from the name.
also: we could take a `std::filesystem::path` to indicate clearly that we expect a path.
```suggestion
std::unique_ptr<Options> Options::from_json_file(const std::string& options_json_path){
```

File: source/Options.cpp [hunk 20]
```diff
@@ -147,336 +147,164 @@ Options::Options(
       export_origins_mode_(export_origins_mode),
       propagate_across_arguments_(propagate_across_arguments) {}
 
-Options::Options(const boost::program_options::variables_map& variables) {
-  system_jar_paths_ = parse_paths_list(
-      variables["system-jar-paths"].as<std::string>(),
-      std::nullopt,
-      /* check exist */ false);
+Options::Options(const Json::Value &json){
+    system_jar_paths_ = JsonValidation::string_list(json,"system-jar-paths");
```
Comment (arthaud) [inline source/Options.cpp:151]: I forgot we had this `parse_paths_list` function with all the extra logic.
it seems like you decided to keep these for now, except for `system_jar_paths_`.

I think in the future we should move the logic of `parse_paths_list` into the shim, but that seems out of the scope of this PR.

for now, can we keep `system-jar-paths` as a string and use `parse_paths_list` logic here?

**Review by arthaud — CHANGES_REQUESTED** [review]
Nice, that looks very good!
Just a last change and we will be merging this.
Thanks for all your work.

**Review by arthaud — APPROVED** [review]
